src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -629,7 +629,11 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   def typedMatch(tree: untpd.Match, pt: Type)(implicit ctx: Context) = track("typedMatch") {
     tree.selector match {
       case EmptyTree =>
-        typed(desugar.makeCaseLambda(tree.cases) withPos tree.pos, pt)
+        val arity = pt match {
+          case defn.FunctionType(args, _) => args.length
+          case _ => 1
+        }
```

Comment:
Recent versions of scalac (2.10+?) are more lenient, the expected type of a patmat anon function need only have a `FunctionN` with fully defined argument types as base type.

```
class Test {
  trait X extends Function1[Int, String]
  implicit def f2x(f: Function1[Int, String]): X = ???
  ({case _ if "".isEmpty => ""} : X) // allowed, implicit view used to adapt

  ({case _ if "".isEmpty => 0} : X) // expected String, found Int
}
```

This change was motivated by spores.